File: samples/da-SnowWizard/SnowWizard/src/services/snow_incidents.ts
//Uses the ServiceNow Rest API to deal with incidents
//Author: crisag@microsoft.com

import axios from 'axios';
import * as dotenv from 'dotenv';

dotenv.config({ path: 'env/.env.local.user' });

class IncidentsApiService {
    
    private SN_INSTANCE: string;
    private SN_USERNAME: string;
    private SN_PASSWORD: string;

    constructor() {
        // Environment variables setup
        this.SN_INSTANCE = process.env.SN_INSTANCE || '';
        this.SN_USERNAME = process.env.SN_USERNAME || '';
        this.SN_PASSWORD = process.env.SN_PASSWORD || '';
    }

    // Function to fetch a single incident from ServiceNow
    async getIncident(id: string) {
    try {
        const response = await axios.get(
        `https://${this.SN_INSTANCE}.service-now.com/api/now/table/incident`,
        {
            params: {
            sysparm_limit: 1,
            sysparm_fields: 'number,made_sla,short_description,description,priority,opened_at',
            sysparm_query: `number=${id}`
            },
            auth: {
            username: this.SN_USERNAME,
            password: this.SN_PASSWORD
            },
            headers: {
            'Content-Type': 'application/json',
            },
        }
        );
        console.log('Incidents fetched successfully from ServiceNow:', response.data.result);
        // Extracting incidents from response
        return response.data.result;
    } catch (error) {
        console.error('Error fetching incidents:', error);
        throw error;
    }
    }

    // Function to fetch the latest 10 incidents from ServiceNow
    async getIncidents() {
    try {
        const response = await axios.get(
        `https://${this.SN_INSTANCE}.service-now.com/api/now/table/incident`,
        {
            params: {
            sysparm_limit: 10,
            sysparm_fields: 'number,made_sla,short_description,description,priority,opened_at',
            sysparm_query: 'ORDERBYDESCsys_created_on'
            },
            auth: {
            username: this.SN_USERNAME,
            password: this.SN_PASSWORD
            },
            headers: {
            'Content-Type': 'application/json',
            },
        }
        );
        console.log('Incidents fetched successfully from ServiceNow:', response.data.result);
        // Extracting incidents from response
        return response.data.result;
    } catch (error) {
        console.error('Error fetching incidents:', error);
        throw error;
    }
    }

    // Function to fetch the latest 10 incidents from ServiceNow
    async getUserIncidents(username: string) {
        try {
            //get the sys_id of the user
            const sys_id = await this.getUserSysId(username);
            //fetch incidents assigned to the user
            const response = await axios.get(
            `https://${this.SN_INSTANCE}.service-now.com/api/now/table/incident`,
            {
                params: {
                sysparm_limit: 10,
                sysparm_fields: 'number,made_sla,short_description,description,priority,opened_at',
                sysparm_query: `ORDERBYDESCsys_created_on^assigned_to=${sys_id}`
                },
                auth: {
                username: this.SN_USERNAME,
                password: this.SN_PASSWORD
                },
                headers: {
                'Content-Type': 'application/json',
                },
            }
            );
            console.log('Incidents fetched successfully from ServiceNow:', response.data.result);
            // Extracting incidents from response
            return response.data.result;
        } catch (error) {
            console.error('Error fetching incidents:', error);
            throw error;
        }
        }

        // Function to create a new incident in ServiceNow
        async createIncident(email: string, short_description: string, description: string) {
            try {
                //get the sys_id of the user
                const sys_id = await this.getUserSysId(email);
                // Create a new incident on Service Now
                const response = await axios.post(
                `https://${this.SN_INSTANCE}.service-now.com/api/now/table/incident`,
                {
                    short_description: `${short_description}`,
                    description: `${description}`,
                    caller_id: sys_id
                },
                {
                    auth: {
                    username: this.SN_USERNAME,
                    password: this.SN_PASSWORD
                    },
                    headers: {
                    'Content-Type': 'application/json',
                    },
                }
                );
                console.log('Incident created successfully in ServiceNow:', response.data.result);
                // Extracting incident from response
                return response.data.result;
            } catch (error) {
                console.error('Error creating incident:', error);
                throw error;
            }
        }

        private async getUserSysId(username: string) {
            const response = await axios.get(
                `https://${this.SN_INSTANCE}.service-now.com/api/now/table/sys_user`,
                {
                    params: {
                    sysparm_limit: 10,
                    sysparm_query: `email=${username}`
                    },
                    auth: {
                    username: this.SN_USERNAME,
                    password: this.SN_PASSWORD
                    },
                    headers: {
                    'Content-Type': 'application/json',
                    },
                }
                );
                console.log('User fetched successfully from ServiceNow:', response.data.result);
                // Extracting user sys_id from response
                return response.data.result[0].sys_id;
        }

}

export default new IncidentsApiService();

File: samples/da-snowwizard-pytho/services/snow_incidents.py
# Uses the ServiceNow Rest API to deal with incidents

import os
import requests
from dotenv import load_dotenv

load_dotenv("env/.env.local.user")


class IncidentsApiService:

    def __init__(self):
        self.sn_instance = os.environ.get("SN_INSTANCE", "")
        self.sn_username = os.environ.get("SN_USERNAME", "")
        self.sn_password = os.environ.get("SN_PASSWORD", "")

    @property
    def _base_url(self) -> str:
        return f"https://{self.sn_instance}.service-now.com/api/now/table"

    @property
    def _auth(self) -> tuple:
        return (self.sn_username, self.sn_password)

    @property
    def _headers(self) -> dict:
        return {"Content-Type": "application/json"}

    def get_incident(self, incident_id: str) -> list:
        """Fetch a single incident from ServiceNow by number."""
        response = requests.get(
            f"{self._base_url}/incident",
            params={
                "sysparm_limit": 1,
                "sysparm_fields": "number,made_sla,short_description,description,priority,opened_at",
                "sysparm_query": f"number={incident_id}",
            },
            auth=self._auth,
            headers=self._headers,
        )
        response.raise_for_status()
        result = response.json().get("result", [])
        print(f"Incidents fetched successfully from ServiceNow: {result}")
        return result

    def get_incidents(self) -> list:
        """Fetch the latest 10 incidents from ServiceNow."""
        response = requests.get(
            f"{self._base_url}/incident",
            params={
                "sysparm_limit": 10,
                "sysparm_fields": "number,made_sla,short_description,description,priority,opened_at",
                "sysparm_query": "ORDERBYDESCsys_created_on",
            },
            auth=self._auth,
            headers=self._headers,
        )
        response.raise_for_status()
        result = response.json().get("result", [])
        print(f"Incidents fetched successfully from ServiceNow: {result}")
        return result

    def get_user_incidents(self, username: str) -> list:
        """Fetch the latest 10 incidents assigned to a user."""
        sys_id = self._get_user_sys_id(username)
        response = requests.get(
            f"{self._base_url}/incident",
            params={
                "sysparm_limit": 10,
                "sysparm_fields": "number,made_sla,short_description,description,priority,opened_at",
                "sysparm_query": f"ORDERBYDESCsys_created_on^assigned_to={sys_id}",
            },
            auth=self._auth,
            headers=self._headers,
        )
        response.raise_for_status()
        result = response.json().get("result", [])
        print(f"Incidents fetched successfully from ServiceNow: {result}")
        return result

    def create_incident(self, email: str, short_description: str, description: str) -> dict:
        """Create a new incident in ServiceNow."""
        sys_id = self._get_user_sys_id(email)
        response = requests.post(
            f"{self._base_url}/incident",
            json={
                "short_description": short_description,
                "description": description,
                "caller_id": sys_id,
            },
            auth=self._auth,
            headers=self._headers,
        )
        response.raise_for_status()
        result = response.json().get("result", {})
        print(f"Incident created successfully in ServiceNow: {result}")
        return result

    def _get_user_sys_id(self, username: str) -> str:
        """Get the sys_id of a user by email."""
        response = requests.get(
            f"{self._base_url}/sys_user",
            params={
                "sysparm_limit": 10,
                "sysparm_query": f"email={username}",
            },
            auth=self._auth,
            headers=self._headers,
        )
        response.raise_for_status()
        result = response.json().get("result", [])
        print(f"User fetched successfully from ServiceNow: {result}")
        return result[0]["sys_id"]
